lib/f_core/net/application/c_tftp_server_tenant.cpp

```diff
@@ -0,0 +1,182 @@
+#include "f_core/net/application/c_tftp_server_tenant.h"
+#include "f_core/os/c_file.h"
+
+#include <cstdio>
+#include <cstring>
+#include <zephyr/fs/fs.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(CTftpServerTenant);
+
+char *inet_ntoa(struct in_addr in)
+{
+    static char buf[INET_ADDRSTRLEN];
+    unsigned char *bytes = (unsigned char *)&in.s_addr;
+
+    snprintf(buf, sizeof(buf), "%d.%d.%d.%d", bytes[0], bytes[1], bytes[2], bytes[3]);
+
+    return buf;
+}
+
+void CTftpServerTenant::Startup() {
+}
+
+void CTftpServerTenant::Cleanup() {
+}
+
+void CTftpServerTenant::Run() {
+    sockaddr srcAddr = {0};
+    socklen_t srcAddrLen = sizeof(srcAddr);
+    uint8_t packet[rwRequestPacketSize] = {0};
+    int rxLen = sock.ReceiveAsynchronous(packet, rwRequestPacketSize, &srcAddr, &srcAddrLen);
+
+    if (rxLen == 0) {
+        return;
+    } else if (rxLen < 0) {
+        LOG_ERR("Failed to receive packet (%d)", rxLen);
+        return;
+    }
+
+    // Opcode is the first two bytes in network order.
+    uint16_t opcode = (packet[0] << 8) | packet[1];
+    switch (opcode) {
+        case RRQ:
+            handleReadRequest(srcAddr, packet, rxLen);
+            break;
+        default:
+            LOG_ERR("Undefined opcode operation %d", opcode);
+    }
+}
+
+int CTftpServerTenant::waitForAck(CUdpSocket &dataSock, const sockaddr &srcAddr, uint16_t blockNum) {
+    static constexpr uint16_t ACK_PKT_SIZE = 4;
+    uint8_t ack[ACK_PKT_SIZE] = {0}; // 2 bytes for opcode, 2 for block number
+    const int maxRetries = 5;
+    int retries = 0;
+
+    while (retries < maxRetries) {
+        int ret = dataSock.ReceiveAsynchronous(ack, 4, const_cast<sockaddr *>(&srcAddr), nullptr);
+        if (ret < 0) {
+            // If no data available yet, wait and retry.
+            if (errno == EWOULDBLOCK || errno == EAGAIN) {
+                k_sleep(K_MSEC(100));
+                retries++;
+                continue;
+            } else {
+                LOG_ERR("Failed to receive ACK (%d)", errno);
+                return errno;
+            }
+        } else if (ret == ACK_PKT_SIZE) {
+            uint16_t receivedOpcode = (ack[0] << 8) | ack[1];
+            uint16_t receivedBlockNum = (ack[2] << 8) | ack[3];
+
+            if (receivedOpcode == ACK && receivedBlockNum == blockNum) {
+                return 0; // Valid ACK received.
+            }
+
+            LOG_ERR("Received invalid ACK. Expected block %d, received block %d", blockNum, receivedBlockNum);
+            return -1;
+        }
+    }
+
+    LOG_ERR("Timeout waiting for ACK for block %d", blockNum);
+    return -1;
+}
+
+void CTftpServerTenant::handleReadRequest(const sockaddr &clientAddr, const uint8_t *packet, int len) {
+    const char *filename = reinterpret_cast<const char *>(&packet[2]);
+    const char *modeStr = reinterpret_cast<const char *>(&packet[2 + strlen(filename) + 1]);
```

Comment:
bros strlen ing a string from the network. we're getting hacked

Reply:
I wanted to strnlen, but then I was like who is gonna send garbage tftp requests to the flight computer they need to be wired into x)